src/EFCore.PG/Metadata/PostgresRange.cs

```diff
@@ -0,0 +1,195 @@
+        }
+
+        [NotNull]
+        static string BuildAnnotationName(string schema, string name)
+            => NpgsqlAnnotationNames.RangePrefix + (schema == null ? name : schema + '.' + name);
+
+        [NotNull]
+        public static IEnumerable<PostgresRange> GetPostgresRanges([NotNull] IAnnotatable annotatable)
+            => Check.NotNull(annotatable, nameof(annotatable))
+                .GetAnnotations()
+                .Where(a => a.Name.StartsWith(NpgsqlAnnotationNames.RangePrefix, StringComparison.Ordinal))
+                .Select(a => new PostgresRange(annotatable, a.Name));
+
+        [NotNull]
+        public Annotatable Annotatable => (Annotatable)_annotatable;
+
+        [NotNull]
+        public string Schema => GetData().Schema;
+
+        [NotNull]
+        public string Name => GetData().Name;
+
+        [NotNull]
+        public string Subtype
+        {
+            get => GetData().Subtype;
+            set
+            {
+                var x = GetData();
+                var (_, _, _, canonicalFunction, subtypeOpClass, collation, subtypeDiff) = GetData();
+                SetData(value, canonicalFunction, subtypeOpClass, collation, subtypeDiff);
+            }
+        }
+
+        [CanBeNull]
+        public string CanonicalFunction
+        {
+            get => GetData().CanonicalFunction;
+            set
+            {
+                var x = GetData();
+                var (_, _, subtype, _, subtypeOpClass, collation, subtypeDiff) = GetData();
+                SetData(subtype, value, subtypeOpClass, collation, subtypeDiff);
+            }
+        }
+
+        [CanBeNull]
+        public string SubtypeOpClass
+        {
+            get => GetData().SubtypeOpClass;
+            set
+            {
+                var x = GetData();
+                var (_, _, subtype, canonicalFunction, _, collation, subtypeDiff) = GetData();
+                SetData(subtype, canonicalFunction, value, collation, subtypeDiff);
+            }
+        }
+
+        [CanBeNull]
+        public string Collation
+        {
+            get => GetData().Collation;
+            set
+            {
+                var x = GetData();
+                var (_, _, subtype, canonicalFunction, subtypeOpClass, _, subtypeDiff) = GetData();
+                SetData(subtype, canonicalFunction, subtypeOpClass, value, subtypeDiff);
+            }
+        }
+
+        [CanBeNull]
+        public string SubtypeDiff
+        {
+            get => GetData().SubtypeDiff;
+            set
+            {
+                var x = GetData();
+                var (_, _, subtype, canonicalFunction, subtypeOpClass, collation, _) = GetData();
+                SetData(subtype, canonicalFunction, subtypeOpClass, collation, value);
+            }
+        }
+
+        (string Schema, string Name, string Subtype, string CanonicalFunction, string SubtypeOpClass, string Collation,
+            string SubtypeDiff) GetData()
```

Comment:
nit: partially chopped return type.

This might be a wide enough tuple to justify defining a full type... But not a big deal either way.

Reply:
PostgresRange itself already represents more or less the tuple of all these variables, with the capability to serialize/deserialize itself to and from annotations... It's not the prettiest, I agree.

At some point we can look at refactor how all our annotations are implemented, but for now  let's continue in the way others are done.